### ISSUE TITLE
How can I use focal loss to train RPN? 

### ISSUE THREAD

**Issue by derek-zr** [facebookresearch/Detectron#439]
I want to try to use focal loss to train RPN (not RetinaNet, but with Mask RCNN) . But when I modify the code, I meet problems.
The code is in detectron/modeling/FPN.py  

def add_fpn_rpn_losses(model):

    """Add RPN on FPN specific losses."""

    loss_gradients = {}

    for lvl in range(cfg.FPN.RPN_MIN_LEVEL, cfg.FPN.RPN_MAX_LEVEL + 1):

        slvl = str(lvl)

        # Spatially narrow the full-sized RPN label arrays to match the feature map

        # shape

        model.net.SpatialNarrowAs(

            ['rpn_labels_int32_wide_fpn' + slvl, 'rpn_cls_logits_fpn' + slvl],

            'rpn_labels_int32_fpn' + slvl

        )
        for key in ('targets', 'inside_weights', 'outside_weights'):

            model.net.SpatialNarrowAs(

                [

                    'rpn_bbox_' + key + '_wide_fpn' + slvl,

                    'rpn_bbox_pred_fpn' + slvl

                ],

                'rpn_bbox_' + key + '_fpn' + slvl

            )
        loss_rpn_cls_fpn = model.net.SigmoidFocalLoss(

            ['rpn_cls_logits_fpn' + slvl,

                'rpn_labels_int32_fpn' + slvl],

            'loss_rpn_cls_fpn' + slvl,

            gamma=cfg.RETINANET.LOSS_GAMMA,

            alpha=cfg.RETINANET.LOSS_ALPHA,

            scale=(

                model.GetLossScale() / cfg.TRAIN.RPN_BATCH_SIZE_PER_IM /

                cfg.TRAIN.IMS_PER_BATCH

            ),

            num_classes=model.num_classes - 1

        )

        # Normalization by (1) RPN_BATCH_SIZE_PER_IM and (2) IMS_PER_BATCH is

        # handled by (1) setting bbox outside weights and (2) SmoothL1Loss

        # normalizes by IMS_PER_BATCH

        loss_rpn_bbox_fpn = model.net.SelectSmoothL1Loss(

            [

                'rpn_bbox_pred_fpn' + slvl,

                'rpn_bbox_targets_fpn' + slvl,

                'rpn_bbox_inside_weights_fpn' + slvl,

                'rpn_bbox_outside_weights_fpn' + slvl

            ],

            'loss_rpn_bbox_fpn' + slvl,

            beta=cfg.RETINANET.BBOX_REG_BETA,

            scale=model.GetLossScale() * cfg.RETINANET.BBOX_REG_WEIGHT,

        )
        loss_gradients.update(

            blob_utils.get_loss_gradients(model, [loss_rpn_cls_fpn, loss_rpn_bbox_fpn])
        )

        model.AddLosses(['loss_rpn_cls_fpn' + slvl,

                         'loss_rpn_bbox_fpn' + slvl])

    return loss_gradients

But when I run the train_net.py, I got the error:
![image](https://user-images.githubusercontent.com/28928339/40277490-4cfb04b6-5c52-11e8-8c23-695951d26234.png)

Have anyone tried to use focal loss to train RPN ? Can you give me some suggestions about this?

**Comment by withyou1771** [comment]
I add focalloss to fpn in my project
[](https://github.com/withyou1771/Detectron_FocalLoss)
Welcome to discuss with me

**Comment by withyou1771** [comment]
I add focalloss to fpn in my project
[click here](https://github.com/withyou1771/Detectron_FocalLoss)
Welcome to discuss with me

**Comment by withyou1771** [comment]
I add focalloss to fpn in my project but no work
[click here](https://github.com/withyou1771/Detectron_FocalLoss)
Welcome to discuss with me

**Comment by withyou1771** [comment]
I add focalloss to fpn in my project but no work
[click here](https://github.com/withyou1771/Detectron_FocalLoss)
I‘m not sure how to set the input blob of model.net.SigmoidFocalLoss.
 ['rpn_cls_logits_fpn' + slvl, 'rpn_labels_int32_fpn' + slvl]
It should be 3 parameters instead of 2.
Welcome to discuss with me

**Comment by derek-zr** [comment]
I meet the same problem. I see the caffe2/module/detectron, and find the definition of focal loss op. The input must be 3 parameters. I don't know how to solve this, and set the input properly.

**Comment by zhuliqian** [comment]
I also want to use focal loss,do you solve it?

**Comment by shuoshuoa** [comment]
@derek-zr hello, how did you find the definition of focalloss, can you help me to find the source code?Thank you!!!

**Comment by ash1995** [comment]
@derek-zr Hey, were you able to resolve this issue?

**Comment by derek-zr** [comment]
> @derek-zr Hey, were you able to resolve this issue?

Use another pytorch version to try this. But the result is not good, cause two stage detector already limit the ration between backgroud and foreground. See Libra RCNN, you can see the explanation.

**Comment by derek-zr** [comment]
> @derek-zr hello, how did you find the definition of focalloss, can you help me to find the source code?Thank you!!!

I remember it's in caffe2 source code, maybe in caffe2/detectron/operators ? Just find the definition in your IDE, I think you can find it.

**Comment by derek-zr** [comment]
Closed due to I have already try it in Pytorch. And the result is not good.
